core/src/main/java/org/infinispan/util/concurrent/CommandAckCollector.java

```diff
@@ -258,12 +242,31 @@ private BaseAckTarget(long id, int topologyId) {
          this.timeoutTask = timeoutExecutor.schedule(this, timeoutNanoSeconds, TimeUnit.NANOSECONDS);
       }
 
+      /**
+       * Invoked when the future is completed, it must cleanup all task related to this collector.
```

Comment:
Since the future is not `timeoutTask`, the comment is a bit ambiguous. I suggest adding a protected method `cleanupAfterFuture(CF)` and call that from the subclasses so you can reference that in the javadoc and so subclasses don't have to know whether it's a `BiConsumer` or a `BiFunction`.

I also suggest extracting the entire body in a separate method, e.g. `cleanup()`: with a name like that, it wouldn't need as much javadoc.

Reply:
With the ResponseCollector approach I'll revert the BiConsumer / BiFunction change, so this won't be needed.